src/euphorie/content/browser/similar_titles.py

```diff
@@ -0,0 +1,181 @@
+from collections import defaultdict
+from euphorie.content import MessageFactory as _
+from euphorie.content.surveygroup import ISurveyGroup
+from nltk.corpus import stopwords
+from nltk.stem import WordNetLemmatizer
+from nltk.tokenize import word_tokenize
+from plone import api
+from plone.autoform.form import AutoExtensibleForm
+from plone.memoize import ram
+from plone.memoize.view import memoize
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.metrics.pairwise import cosine_similarity
+from z3c.form import button
+from z3c.form import form
+from zope import schema
+from zope.interface import Interface
+
+
+def _forever_cache_key(method, *args, **kwargs):
+    return method.__name__
+
+
+class ISimilarTitleSchema(Interface):
+    min_similarity = schema.Float(
+        title=_("Minimum similarity"),
+        description=_("Minimum similarity between titles"),
+        default=0.5,
+        required=True,
+        min=0.0,
+        max=1.0,
+    )
+
+    max_similarity = schema.Float(
+        title=_("Maximum similarity"),
+        description=_("Maximum similarity between titles"),
+        default=1.0,
+        required=True,
+        min=0.0,
+        max=1.0,
+    )
+
+
+class SimilarTitles(AutoExtensibleForm, form.Form):
+    """The upload view for a :obj:`euphorie.content.sector`
+
+    View name: @@upload
+    """
+
+    schema = ISimilarTitleSchema
+    ignoreContext = True
+    form_name = _("Similar Titles OiRA Tool")
+
+    label = _("Similar Titles OiRA Tool")
+    description = _("This tool allows you to find objects with similar titles.")
+
+    @property
+    def template(self):
+        return self.index
+
+    def redirect(self, target=None, message=None, message_type="info"):
+        if target is None:
+            target = self.context.absolute_url()
+        if message is not None:
+            api.portal.show_message(
+                message=message, request=self.request, type=message_type
+            )
+        self.request.response.redirect(target)
+
+    @property
+    @memoize
+    def extracted_data(self):
+        """extract_data returns data and errors from the form.
+        We want just the data
+        """
+        return self.extractData()[0]
+
+    @property
+    def min_similarity(self):
+        return self.extracted_data.get("min_similarity", 0.5)
+
+    @property
+    def max_similarity(self):
+        return self.extracted_data.get("max_similarity", 1.0)
+
+    @property
+    def tool_cache(self):
+        return {}
```

Comment:
I don't think I get it... Will this not return a new dict on every call? Why make it an instance method then?

Reply:
I think this misses a `memoize`. Isn't this already a

Reply:
Yeah, that was my question: Why did you make it an instance method? I.e. why not explicitly use `{}` in the method below.

... but now it makes sense! :+1: